sw/device/tests/dma_inline_hashing.c

```diff
@@ -0,0 +1,171 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/status.h"
+#include "sw/ip/base/dif/dif_base.h"
+#include "sw/ip/dma/dif/dif_dma.h"
+#include "sw/ip/pinmux/dif/dif_pinmux.h"
+#include "sw/ip/pinmux/test/utils/pinmux_testutils.h"
+#include "sw/ip/rv_core_ibex/test/utils/rand_testutils.h"
+#include "sw/ip/spi_host/dif/dif_spi_host.h"
+#include "sw/lib/sw/device/arch/device.h"
+#include "sw/lib/sw/device/base/mmio.h"
+#include "sw/lib/sw/device/runtime/hart.h"
+#include "sw/lib/sw/device/runtime/log.h"
+
+#include "hw/top_darjeeling/sw/autogen/top_darjeeling.h"
+#include "spi_host_regs.h"  // Generated.
+
+// The TX_SIZE must be in sync with the data size in spi_device_dma_seq.sv
+// 1 SPI segment can only transfer at maximum 512 bytes
+#define TX_SIZE 512
+#define CHUNK_SIZE 32 * 4  // Half the SPI host FIFO size
```

Comment:
Should we also test transfers of multiple SPI segments, to ensure that hashing works correctly with multiple DMA transfers (i.e., when the `initial_transfer` bit is set for the first transfer but not for subsequent transfers, which also contribute to the hashing)? I think this would be relevant for the firmware measurement use case

Reply:
It already does a transfer of 512 bytes, chunked at 128 bytes. So, it performs 4 segments, with `initial_transfer` set only on the first segement.

Reply:
Hm, `dif_dma_start` always sets `DMA_CONTROL_INITIAL_TRANSFER_BIT` to 1, though. Maybe this is only relevant for non-hardware-handshake mode, though?

Reply:
Probably the DIF should be changed with a default parameter for the initial transfer bit. This however only applies to non-hardware handshake transfers, where multiple, individual transfers (`dif_dma_start` is asserted multiple times) can contribute to the same hash. In hardware handshake `dif_dma_start` is asserted only once. During that transfers, multiple hardware triggers are asserted, indicating that a new segment is available at the LSIO device.

Reply:
I see, thanks for the explanation :+1: